Type mismatch in function arguments E0631, E0271 are falsely recognized as E0308 mismatched types

Regression from `1.45.0-beta.4 (2020-07-13 e99e6422a2f799fc86a3)` to `1.46.0-nightly (2020-07-15 7e11379f3b4c376fbb9a)`.

The following code yields an incorrect error:
```rust
use std::convert::identity;

pub fn is_sorted_and_has_duplicates_by<T, S: PartialEq + PartialOrd>(
    data: &[T],
    key: impl Fn(&T) -> S,
) -> Option<&T> {
    unimplemented!()
}

pub fn is_sorted_and_has_duplicates<T: PartialEq + PartialOrd>(data: &[T]) -> Option<&T> {
    is_sorted_and_has_duplicates_by(data, identity)
}
```
On beta, this gives:
```
error[E0631]: type mismatch in function arguments
  --> src/lib.rs:11:43
   |
3  | pub fn is_sorted_and_has_duplicates_by<T, S: PartialEq + PartialOrd>(
   |        ------------------------------- required by a bound in this
4  |     data: &[T],
5  |     key: impl Fn(&T) -> S,
   |               ----------- required by this bound in `is_sorted_and_has_duplicates_by`
...
11 |     is_sorted_and_has_duplicates_by(data, identity)
   |                                           ^^^^^^^^
   |                                           |
   |                                           expected signature of `for<'r> fn(&'r T) -> _`
   |                                           found signature of `fn(_) -> _`

error[E0271]: type mismatch resolving `for<'r> <fn(_) -> _ {std::convert::identity::<_>} as std::ops::FnOnce<(&'r T,)>>::Output == _`
  --> src/lib.rs:11:5
   |
3  | pub fn is_sorted_and_has_duplicates_by<T, S: PartialEq + PartialOrd>(
   |        ------------------------------- required by a bound in this
4  |     data: &[T],
5  |     key: impl Fn(&T) -> S,
   |                         - required by this bound in `is_sorted_and_has_duplicates_by`
...
11 |     is_sorted_and_has_duplicates_by(data, identity)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected bound lifetime parameter, found concrete lifetime
```

On nightly, this:
```
error[E0308]: mismatched types
  --> src/lib.rs:11:5
   |
11 |     is_sorted_and_has_duplicates_by(data, identity)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
   |
   = note: expected type `std::ops::FnOnce<(&T,)>`
              found type `std::ops::FnOnce<(&T,)>`
```

--- a/src/test/ui/lifetimes/lifetime-errors/issue_74400.nll.stderr
+++ b/src/test/ui/lifetimes/lifetime-errors/issue_74400.nll.stderr
@@ -1,0 +1,30 @@
+error[E0310]: the parameter type `T` may not live long enough
+  --> $DIR/issue_74400.rs:12:5
+   |
+LL |     f(data, identity)
+   |     ^^^^^^^^^^^^^^^^^
+   |
+   = help: consider adding an explicit lifetime bound `T: 'static`...
+
+error[E0308]: mismatched types
+  --> $DIR/issue_74400.rs:12:5
+   |
+LL |     f(data, identity)
+   |     ^^^^^^^^^^^^^^^^^ one type is more general than the other
+   |
+   = note: expected type `for<'r> Fn<(&'r T,)>`
+              found type `Fn<(&T,)>`
+
+error: implementation of `FnOnce` is not general enough
+  --> $DIR/issue_74400.rs:12:5
+   |
+LL |     f(data, identity)
+   |     ^^^^^^^^^^^^^^^^^ implementation of `FnOnce` is not general enough
+   |
+   = note: `fn(&'2 T) -> &'2 T {identity::<&'2 T>}` must implement `FnOnce<(&'1 T,)>`, for any lifetime `'1`...
+   = note: ...but it actually implements `FnOnce<(&'2 T,)>`, for some specific lifetime `'2`
+
+error: aborting due to 3 previous errors
+
+Some errors have detailed explanations: E0308, E0310.
+For more information about an error, try `rustc --explain E0308`.

--- a/src/test/ui/lifetimes/lifetime-errors/issue_74400.rs
+++ b/src/test/ui/lifetimes/lifetime-errors/issue_74400.rs
@@ -1,0 +1,13 @@
+//! Regression test for #74400: Type mismatch in function arguments E0631, E0271 are falsely
+//! recognized as E0308 mismatched types.
+
+use std::convert::identity;
+
+fn main() {}
+
+fn f<T, S>(data: &[T], key: impl Fn(&T) -> S) {
+}
+
+fn g<T>(data: &[T]) {
+    f(data, identity) //~ ERROR implementation of `FnOnce` is not general
+}

--- a/src/test/ui/lifetimes/lifetime-errors/issue_74400.stderr
+++ b/src/test/ui/lifetimes/lifetime-errors/issue_74400.stderr
@@ -1,0 +1,11 @@
+error: implementation of `FnOnce` is not general enough
+  --> $DIR/issue_74400.rs:12:5
+   |
+LL |     f(data, identity)
+   |     ^ implementation of `FnOnce` is not general enough
+   |
+   = note: `fn(&'2 T) -> &'2 T {identity::<&'2 T>}` must implement `FnOnce<(&'1 T,)>`, for any lifetime `'1`...
+   = note: ...but it actually implements `FnOnce<(&'2 T,)>`, for some specific lifetime `'2`
+
+error: aborting due to previous error
+

